server/pxf-hdfs/src/main/java/org/greenplum/pxf/plugins/hdfs/ParquetResolver.java

```
@@ -257,10 +262,37 @@ private void fillGroupWithPrimitive(Group group, int columnIndex, Object fieldVa
+        if (!NumberUtils.isNumber(value)) {
+            throw new UnsupportedTypeException(String.format("Invalid numeric %s", value));
+        }
+
         // From org.apache.hadoop.hive.ql.io.parquet.write.DataWritableWriter.DecimalDataWriter#decimalToBinary
         DecimalLogicalTypeAnnotation typeAnnotation = (DecimalLogicalTypeAnnotation) type.getLogicalTypeAnnotation();
         int precision = Math.min(HiveDecimal.MAX_PRECISION, typeAnnotation.getPrecision());
         int scale = Math.min(HiveDecimal.MAX_SCALE, typeAnnotation.getScale());
+
+        BigDecimal bigDecimal = NumberUtils.createBigDecimal(value);
+        int dataPrecision = bigDecimal.precision();
+        int dataScale = bigDecimal.scale();
+
+        String decimalOverflowOption = parseDecimalOverflowOption(configuration);
+        // data precision overflow and decimal overflow option is error
+        if (dataPrecision > HiveDecimal.MAX_PRECISION) {
+            if (decimalOverflowOption.equals(ParquetWriteDecimalOverflowOption.ERROR.getValue())) {
+                throw new UnsupportedTypeException(String.format("Data size of data %s exceeds the maximum numeric precision %d.",
+                        value, HiveDecimal.MAX_PRECISION));
+            }
+            LOG.warn("Data precision overflow. Overflowing data wiil be set to empty. Query will success but data is incomplete");
+        }
+
+        int integerDigitCount = dataPrecision - dataScale;
+        int maxIntegerDigitCount = precision - scale;
+        //  data integer digit count overflow and decimal overflow option is error
+        if (integerDigitCount > maxIntegerDigitCount && decimalOverflowOption.equals(ParquetWriteDecimalOverflowOption.ERROR.getValue())) {
+            throw new UnsupportedTypeException(String.format("Integer digit count of data %s overflows. " +
+                    "Your integer digit count is %d. The max integer digit count is %d", value, integerDigitCount, maxIntegerDigitCount));
+        }
+
```

why all this logic above ? I thought we would call `HiveDecimal.create(value)` and if the incoming data was not null but we got back NULL, then we have a problem and need to throw the exception (conditionally).

---

There is another case that `HiveDecimal.create()` can return null when column is defined as `NUMERIC`. `HiveDecimal.create` mainly to see if the incoming value can fit into the provided precision. 
- If the integer digit count of the incoming value > provided precision, cannot place all the integer digits -> return null
- if the integer digit count of the incoming value <= provided precision
  - if overall data precision > provided precision -> round off the decimal part in the data
  - if overall data precision <= provided precision -> create a HiveDecimal with the same value

And `HiveDecimal.enforcePrecisionScale` mainly to enforce Decimal(precision) into Decimal(precision, scale). Although `HiveDecimal.create` has ensured the data precision, it is still possible that integer digit count is still greater than the maximum supported integer digit count. Then null value is returned in this case.

For example, a column `num` is defined as `NUMERIC`, and pxf will treat it as `DECIMAL(38,18)` in parquet schema. 
- When we want to insert a value like `1234567890123456789012345678901234567890.123`, it has 40 integer digits which is greater than precision 38, HiveDecimal.create will return `null` and this data will be stored as null
- When we want to insert a value like `123456789012345678901234567890.1234567890`, its precision is 40 which is greater than 38, but its integer digit count is less than 38. The value will be rounded off into `123456789012345678901234567890.12345679`. But DECIMAL(38,18) only allow 20 integer digits, that means this value still cannot fit in this requirement. So `HiveDecimal.enforcePrecisionScale` will return null and this value will be stored as null.

Columns defined as `NUMERIC(precision, scale)` don't need to care about these cases since GPDB has ensured the value can fit in `DECIMAL(precision, scale)` when inserting.